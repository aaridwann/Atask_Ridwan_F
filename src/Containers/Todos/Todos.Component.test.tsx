import { render, waitFor, fireEvent } from "@testing-library/react-native";

import TodoComponent from "./Todos.Component";

import type { Props } from "./Todos.Types";

beforeEach(() => {
  jest.clearAllMocks();
  jest.restoreAllMocks();
});

jest
  .mock("@expo/vector-icons/Ionicons", () => "Ionicons")
  .mock("@react-native-async-storage/async-storage", () => {
    return jest.fn().mockReturnValue({ notes: [1, 2, 3] });
  })
  .mock("../../Utils/CollectData/collectData", () =>
    jest.fn().mockResolvedValue({ notes: [1, 2, 3] })
  );

describe("Todo Compoennt", () => {
  const mockNotes = [
    {
      id: "001",
      title: "title 1",
      description: "description 1",
      isMarked: false,
    },
    {
      id: "002",
      title: "title 2",
      description: "description 2",
      isMarked: false,
    },
  ];
  const Props: Props = {
    navigation: {
      push: jest.fn(),
    },
    addNote: jest.fn(),
    editNote: jest.fn(),
    deleteNote: jest.fn(),
    notes: mockNotes,
  };

  it("SnapTesting", async () => {
    let wrapper = render(<TodoComponent {...Props} />);

    await waitFor(() => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  it("Should render modal while add button pressed", async () => {
    const { getByTestId, debug } = render(<TodoComponent {...Props} />);

    const modal = getByTestId("Modal");
    const addButton = getByTestId("Add_Button");

    await waitFor(() => {
      expect(modal.props.visible).toBe(false);
    });

    fireEvent.press(addButton);

    await waitFor(() => {
      expect(modal.props.visible).toBe(true);
      expect(modal).toBeDefined();
    });
  });

  describe("Add Notes", () => {
    it("Should call Add note while save button is pressed", async () => {
      const { getByTestId } = render(<TodoComponent {...Props} />);
      const expectedPayloadSave = {
        description: "Input description satu",
        id: expect.any(String),
        isMarked: false,
        title: "Input title satu",
      };
      fireEvent.press(getByTestId("Add_Button"));

      const titleInput = getByTestId("Title_Input");
      const descriptionInput = getByTestId("Description_Input");

      await waitFor(() => {
        fireEvent.changeText(titleInput, "Input title satu");
        fireEvent.changeText(descriptionInput, "Input description satu");
      });

      await waitFor(() => {
        expect(titleInput.props.value).toBe("Input title satu");
        expect(descriptionInput.props.value).toBe("Input description satu");
      });

      await waitFor(() => fireEvent.press(getByTestId("Save_Button")));

      await waitFor(() =>
        expect(Props.addNote).toHaveBeenCalledWith(expectedPayloadSave)
      );
    });
  });

  describe("Delete Notes", () => {
    it("Should call delete while delete button isPressed", async () => {
      const { getAllByLabelText } = render(<TodoComponent {...Props} />);

      await waitFor(() =>
        fireEvent.press(getAllByLabelText("Note_001_Delete_Button")[0])
      );

      await waitFor(() => expect(Props.deleteNote).toHaveBeenCalledWith("001"));
    });
  });

  describe("Edit Notes", () => {
    it("Should call filled input while edit button is pressed", async () => {
      const { getByLabelText, getByTestId } = render(
        <TodoComponent {...Props} />
      );

      /** Simulate Edit button is pressed */
      await waitFor(() =>
        fireEvent.press(getByLabelText("Note_002_Edit_Button"))
      );

      /** Modal is Showed */
      await waitFor(() =>
        expect(getByTestId("Modal").props.visible).toBe(true)
      );

      const titleInput = getByTestId("Title_Input");
      const descriptionInput = getByTestId("Description_Input");

      /** Refill input with value is gonna be edit */
      await waitFor(() => {
        expect(titleInput.props.value).toBe("title 2");
        expect(descriptionInput.props.value).toBe("description 2");
      });

      /** Simulate retype the input */
      await waitFor(() => {
        fireEvent.changeText(titleInput, "title 3");
        fireEvent.changeText(descriptionInput, "description 3");
      });

      await waitFor(() => {
        expect(titleInput.props.value).toBe("title 3");
        expect(descriptionInput.props.value).toBe("description 3");
      });

      /** Simulate Press Edit button */
      await waitFor(() => fireEvent.press(getByTestId("Save_Button")));

      await waitFor(() =>
        expect(Props.editNote).toHaveBeenCalledWith({
          id: "002",
          title: "title 3",
          description: "description 3",
          isMarked: false,
        })
      );
    });
  });
});
