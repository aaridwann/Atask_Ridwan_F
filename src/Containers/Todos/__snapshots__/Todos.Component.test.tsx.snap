// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo Compoennt SnapTesting 1`] = `
<RCTSafeAreaView
  style={
    {
      "backgroundColor": "#283618",
      "display": "flex",
      "flex": 1,
      "paddingTop": 50,
    }
  }
>
  <Modal
    animationType="none"
    deviceHeight={null}
    deviceWidth={null}
    hardwareAccelerated={false}
    hideModalContentWhileAnimating={false}
    onBackdropPress={[Function]}
    onModalHide={[Function]}
    onModalWillHide={[Function]}
    onModalWillShow={[Function]}
    onRequestClose={[Function]}
    panResponderThreshold={4}
    scrollHorizontal={false}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    statusBarTranslucent={false}
    supportedOrientations={
      [
        "portrait",
        "landscape",
      ]
    }
    swipeThreshold={100}
    testID="Modal"
    transparent={true}
    visible={false}
  />
  <View
    style={
      {
        "alignContent": "center",
        "alignItems": "center",
        "height": "100%",
        "justifyContent": "space-between",
        "width": "100%",
      }
    }
  >
    <RCTScrollView
      contentContainerStyle={
        {
          "height": 999.9999999999999,
        }
      }
      data={
        [
          {
            "description": "description 1",
            "id": "001",
            "isMarked": false,
            "title": "title 1",
          },
          {
            "description": "description 2",
            "id": "002",
            "isMarked": false,
            "title": "title 2",
          },
        ]
      }
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      refreshControl={
        <RefreshControlMock
          onRefresh={[Function]}
          refreshing={false}
        />
      }
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      showsVerticalScrollIndicator={false}
      stickyHeaderIndices={[]}
      viewabilityConfigCallbackPairs={[]}
    >
      <RCTRefreshControl />
      <View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              {
                "marginVertical": 10,
              }
            }
          >
            <View
              style={
                {
                  "backgroundColor": "white",
                  "borderRadius": 8,
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "marginVertical": 8,
                  "minHeight": 130,
                  "overflow": "hidden",
                  "padding": 16,
                  "width": 640,
                }
              }
            >
              <View
                style={
                  {
                    "height": "100%",
                  }
                }
              >
                <Text
                  accessibilityLabel="Note_001"
                  style={
                    {
                      "fontWeight": "700",
                    }
                  }
                >
                  title 1
                </Text>
                <Text
                  style={
                    {
                      "marginTop": 10,
                      "textAlign": "left",
                    }
                  }
                >
                  description 1
                </Text>
              </View>
              <View
                style={
                  {
                    "flexDirection": "row",
                  }
                }
              >
                <View
                  accessibilityLabel="Note_001_Edit_Button"
                  accessibilityState={
                    {
                      "busy": undefined,
                      "checked": undefined,
                      "disabled": undefined,
                      "expanded": undefined,
                      "selected": undefined,
                    }
                  }
                  accessibilityValue={
                    {
                      "max": undefined,
                      "min": undefined,
                      "now": undefined,
                      "text": undefined,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    {
                      "opacity": 1,
                    }
                  }
                >
                  <Ionicons
                    color="gray"
                    name="pencil-outline"
                    size={40}
                    style={
                      {
                        "margin": 4,
                        "marginTop": 20,
                      }
                    }
                  />
                </View>
                <View
                  accessibilityLabel="Note_001_Delete_Button"
                  accessibilityState={
                    {
                      "busy": undefined,
                      "checked": undefined,
                      "disabled": undefined,
                      "expanded": undefined,
                      "selected": undefined,
                    }
                  }
                  accessibilityValue={
                    {
                      "max": undefined,
                      "min": undefined,
                      "now": undefined,
                      "text": undefined,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    {
                      "opacity": 1,
                    }
                  }
                >
                  <Ionicons
                    color="gray"
                    name="trash-outline"
                    size={40}
                    style={
                      {
                        "margin": 4,
                        "marginTop": 20,
                      }
                    }
                  />
                </View>
              </View>
            </View>
          </View>
        </View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              {
                "marginVertical": 10,
              }
            }
          >
            <View
              style={
                {
                  "backgroundColor": "white",
                  "borderRadius": 8,
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "marginVertical": 8,
                  "minHeight": 130,
                  "overflow": "hidden",
                  "padding": 16,
                  "width": 640,
                }
              }
            >
              <View
                style={
                  {
                    "height": "100%",
                  }
                }
              >
                <Text
                  accessibilityLabel="Note_002"
                  style={
                    {
                      "fontWeight": "700",
                    }
                  }
                >
                  title 2
                </Text>
                <Text
                  style={
                    {
                      "marginTop": 10,
                      "textAlign": "left",
                    }
                  }
                >
                  description 2
                </Text>
              </View>
              <View
                style={
                  {
                    "flexDirection": "row",
                  }
                }
              >
                <View
                  accessibilityLabel="Note_002_Edit_Button"
                  accessibilityState={
                    {
                      "busy": undefined,
                      "checked": undefined,
                      "disabled": undefined,
                      "expanded": undefined,
                      "selected": undefined,
                    }
                  }
                  accessibilityValue={
                    {
                      "max": undefined,
                      "min": undefined,
                      "now": undefined,
                      "text": undefined,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    {
                      "opacity": 1,
                    }
                  }
                >
                  <Ionicons
                    color="gray"
                    name="pencil-outline"
                    size={40}
                    style={
                      {
                        "margin": 4,
                        "marginTop": 20,
                      }
                    }
                  />
                </View>
                <View
                  accessibilityLabel="Note_002_Delete_Button"
                  accessibilityState={
                    {
                      "busy": undefined,
                      "checked": undefined,
                      "disabled": undefined,
                      "expanded": undefined,
                      "selected": undefined,
                    }
                  }
                  accessibilityValue={
                    {
                      "max": undefined,
                      "min": undefined,
                      "now": undefined,
                      "text": undefined,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    {
                      "opacity": 1,
                    }
                  }
                >
                  <Ionicons
                    color="gray"
                    name="trash-outline"
                    size={40}
                    style={
                      {
                        "margin": 4,
                        "marginTop": 20,
                      }
                    }
                  />
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
    <View
      style={
        {
          "backgroundColor": "white",
          "height": 120.00000000000001,
          "padding": 10,
          "paddingHorizontal": 16,
          "top": 0,
          "width": "90%",
        }
      }
    >
      <RCTScrollView
        contentContainerStyle={
          {
            "overflow": "scroll",
          }
        }
        showsVerticalScrollIndicator={false}
      >
        <View>
          <Text
            style={
              {
                "fontWeight": "800",
              }
            }
          >
            Result of encrypt
          </Text>
          <Text />
        </View>
      </RCTScrollView>
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "white",
        "borderColor": "#283618",
        "borderRadius": 49.99999999999999,
        "borderWidth": 1,
        "bottom": 36,
        "height": 99.99999999999999,
        "justifyContent": "center",
        "justifyItems": "center",
        "opacity": 1,
        "position": "absolute",
        "right": 36,
        "width": 99.99999999999999,
        "zIndex": 1,
      }
    }
    testID="Add_Button"
  >
    <Ionicons
      color="gray"
      name="add-circle-outline"
      size={70}
    />
  </View>
</RCTSafeAreaView>
`;
